### PR TITLE
support tensor index.

### PR DESCRIPTION
**Pull request by hbwx24**
### PR types

Function optimization
### PR changes

APIs
### 1.整体支持：
image

image

### 2.具体功能点支持：
image
image
image
image
image

### 3. 支持 tensor类型索引，示例如下：
```
array = np.arange(4*3*2).reshape([4, 3, 2])
value = np.arange(12*3).reshape([3, 2, 3, 2])
index = [[0, 0], [3, 1]]

index_t = paddle.to_tensor(index)
index_np = np.array(index)
tt = paddle.to_tensor(array)

plist = paddle.index_select(tt, index_t, axis=0)

nplist = array[index_np]

print(np.array_equal(plist.numpy(), nplist))

```

### CODE REVIEW

File: paddle/fluid/operators/index_select_op.cc [hunk 1]
```diff
@@ -46,19 +46,22 @@ class IndexSelectOp : public framework::OperatorWithKernel {
...
-
-    auto output_dim = framework::vectorize(input_dim);
     if (dim < 0) {
       dim += input_dim.size();
     }
-    output_dim[dim] = index_dim[0];
+
+    std::vector<int64_t> output_dim(input_dim.size() + index_dim.size() - 1);
+
+    for (int i = 0; i < static_cast<int>(output_dim.size()); i++) {
```
Comment (chenwhql) [inline paddle/fluid/operators/index_select_op.cc]: use `size_t i = 0` directly?

File: python/paddle/fluid/dygraph/varbase_patch_methods.py [hunk 2]
```diff
@@ -559,11 +559,29 @@ def contain_tensor(item):
...
+                    return False
+                if isinstance(item, (tuple, list)):
+                    for s in item:
+                        if not _is_list(s):
+                            return False
+                return True
+
+            if isinstance(item, np.ndarray):
+                return True
+            if not isinstance(item, (tuple, list)):
```
Comment (chenwhql) [inline python/paddle/fluid/dygraph/varbase_patch_methods.py]: do we need to support `set`?

File: python/paddle/fluid/tests/unittests/test_index_select_op.py [hunk 3]
```diff
@@ -131,5 +132,99 @@ def test_dygraph_api(self):
...
+        index = [[1, 2], [0, 3]]
+
+        index_p = paddle.to_tensor(index)
+        pt = paddle.to_tensor(array, stop_gradient=False)
+
+        y = pt[index_p]
+        y = y * y
+        loss = y.sum()
+        loss.backward()
+        grad_torch = np.array([[[0., 2.], [4., 6.], [8., 10.]],
```
Comment (chenwhql) [inline python/paddle/fluid/tests/unittests/test_index_select_op.py]: rename this var

File: python/paddle/fluid/variable_index.py [hunk 4]
```diff
@@ -32,7 +32,9 @@ def replace_ellipsis(var, item):
     item = list(item)
 
     # Remove Variable to skip bug when counting Ellipsis
-    item_remove_var = [ele for ele in item if not isinstance(ele, Variable)]
+    item_remove_var = [
```
Comment (chenwhql) [inline python/paddle/fluid/variable_index.py:197]: why need this skip

File: paddle/fluid/operators/index_select_op.cc [hunk 5]
```diff
@@ -46,19 +46,22 @@ class IndexSelectOp : public framework::OperatorWithKernel {
             "to be in range of [-%d, %d]. But received Attr(dim) = %d.",
             input_dim.size(), input_dim.size() - 1, dim));
 
-    PADDLE_ENFORCE_EQ(
```
Comment (chenwhql) [inline paddle/fluid/operators/index_select_op.cc]: 放开这个口子的话，我们是否需要一些别的检查？index_dim会不会有不合理的输入？

File: python/paddle/fluid/variable_index.py [hunk 6]
```diff
@@ -88,6 +90,222 @@ def deal_attrs(attrs, attr, attr_name, tensor_attr_name, inputs, infer_flags):
...
+
+    out_dims_shape[:index_dims_size - 1] = index_dims[1:]
+
+    out_dims_shape[index_dims_size - 1:] = var_dims[index_dims[0]:]
+    return out_dims_shape
+
+
+def index_tensor(tensor, offsets, strides):
+    from . import layers
+    from .framework import Variable
```
Comment (chenwhql) [inline python/paddle/fluid/variable_index.py]: why not import in beginning?

File: python/paddle/fluid/variable_index.py [hunk 7]
```diff
@@ -88,6 +90,222 @@ def deal_attrs(attrs, attr, attr_name, tensor_attr_name, inputs, infer_flags):
...
+            start = int(offsets[i])
+            end = int(offsets[i] + strides[i])
+        tensor_array.append(tensor_flatten[start:end])
+    combine_tensor = layers.stack(tensor_array)
+    return combine_tensor
+
+
+def getitem_list_index(var, list_index):
+    from . import layers
+    from .framework import Variable
```
Comment (chenwhql) [inline python/paddle/fluid/variable_index.py]: same above

File: python/paddle/fluid/variable_index.py [hunk 8]
```diff
@@ -88,6 +90,222 @@ def deal_attrs(attrs, attr, attr_name, tensor_attr_name, inputs, infer_flags):
...
+        offsets[i] = offset
+
+    out_tensor = index_tensor(var, offsets, s)
+    out_tensor = out_tensor.reshape(out_shape)
+
+    return out_tensor
+
+
+def setitem_list_index(var, index_list, value):
+    from . import layers
```
Comment (chenwhql) [inline python/paddle/fluid/variable_index.py]: same above

File: python/paddle/fluid/variable_index.py [hunk 9]
```diff
@@ -151,13 +370,20 @@ def _getitem_impl_(var, item):
 
         elif isinstance(slice_item, list):
             is_bool_list = False
+            is_int_list = False
+            list_index = []
             for i in slice_item:
-                if not isinstance(i, (int, bool)):
-                    raise TypeError("Only support int or bool in index list.")
+                if not isinstance(i, (int, bool, list)):
```
Comment (chenwhql) [inline python/paddle/fluid/variable_index.py]: Isn't it a tuple here?

File: paddle/fluid/operators/index_select_op.cc (hunk 1 above)
Comment (hbwx24) [reply paddle/fluid/operators/index_select_op.cc]: revert the modification of index_select_op.

File: python/paddle/fluid/dygraph/varbase_patch_methods.py (hunk 2 above)
Comment (hbwx24) [reply python/paddle/fluid/dygraph/varbase_patch_methods.py]: revert the modification of index_select_op.

File: python/paddle/fluid/tests/unittests/test_index_select_op.py (hunk 3 above)
Comment (hbwx24) [reply python/paddle/fluid/tests/unittests/test_index_select_op.py]: revert the modification of index_select_op.

File: python/paddle/fluid/variable_index.py (hunk 4 above)
Comment (hbwx24) [reply python/paddle/fluid/variable_index.py:197]: item_remove_var.count(Ellipsis)计算包含Ellipsis的个数，如果对象为Variable或者ndarray，count函数将报错。

File: paddle/fluid/operators/index_select_op.cc (hunk 5 above)
Comment (hbwx24) [reply paddle/fluid/operators/index_select_op.cc]: revert the modification of index_select_op.

File: python/paddle/fluid/variable_index.py (hunk 6 above)
Comment (hbwx24) [reply python/paddle/fluid/variable_index.py]: 改为通过paddle.xx引用。

File: python/paddle/fluid/variable_index.py (hunk 7 above)
Comment (hbwx24) [reply python/paddle/fluid/variable_index.py]: Done, thx.

File: python/paddle/fluid/variable_index.py (hunk 8 above)
Comment (hbwx24) [reply python/paddle/fluid/variable_index.py]: Done, thx.

File: python/paddle/fluid/variable_index.py (hunk 9 above)
Comment (hbwx24) [reply python/paddle/fluid/variable_index.py]: Done, thx.

File: python/paddle/fluid/dygraph/varbase_patch_methods.py [hunk 10]
```diff
@@ -568,6 +586,31 @@ def contain_tensor(item):
...
+                        or isinstance(slice_item.stop, Variable) \
+                           or isinstance(slice_item.step, Variable):
+                        return True
+                else:
+                    if isinstance(slice_item, Variable):
+                        return True
+            return False
+
+        if contain_tensor(item):
+            # 1. Call _getitem_impl_ when item contains tensor.
```
Comment (chenwhql) [inline python/paddle/fluid/dygraph/varbase_patch_methods.py]: get -> set

File: python/paddle/fluid/dygraph/varbase_patch_methods.py [hunk 11]
```diff
@@ -568,6 +586,31 @@ def contain_tensor(item):
...
+            for slice_item in item:
+                if isinstance(slice_item, slice):
+                    if isinstance(slice_item.start, Variable)  \
+                        or isinstance(slice_item.stop, Variable) \
+                           or isinstance(slice_item.step, Variable):
+                        return True
+                else:
+                    if isinstance(slice_item, Variable):
+                        return True
+            return False
```
Comment (zyfncg) [inline python/paddle/fluid/dygraph/varbase_patch_methods.py]: 是否可以和`__getitem__`共用一份`contain_tensor`代码 ?

File: python/paddle/fluid/variable_index.py [hunk 12]
```diff
@@ -16,10 +16,177 @@
...
+
+
+class SliceInfo:
+    def __init__(self):
+        self.pre_shape = None
+        self.indexes = []
+
+    def update(self, index):
+        if is_list_tuple(index, int) or isinstance(
+                index, (paddle.fluid.core.VarBase, paddle.fluid.Variable,
```
Comment (chenwhql) [inline python/paddle/fluid/variable_index.py]: only keep Variable is ok, when in dygraph mode, VarBase is Variable

File: python/paddle/fluid/variable_index.py [hunk 13]
```diff
@@ -16,10 +16,177 @@
...
+                        np.ndarray)):  # Tensor
+            if not isinstance(index, (paddle.fluid.core.VarBase,
```
Comment (chenwhql) [inline python/paddle/fluid/variable_index.py]: same above, fix all other places

File: python/paddle/fluid/variable_index.py [hunk 14]
```diff
@@ -188,35 +366,44 @@ def _getitem_impl_(var, item):
             idx = assign(np.array(slice_item).astype("int32"))
             return index_select(var, index=idx, axis=0)
 
-        elif isinstance(slice_item, Variable):
-            if len(item) != 1:
-                raise IndexError(
-                    "When index contains a Tensor, its length must be 1, but received {}.".
-                    format(len(item)))
+        elif isinstance(slice_item, np.ndarray):
+            # delete
```
Comment (chenwhql) [inline python/paddle/fluid/variable_index.py]: what `delete` mean?

File: python/paddle/fluid/variable_index.py [hunk 15]
```diff
@@ -16,10 +16,177 @@
...
+    def get_offset_stride(self, tensor_shape):
+        for i in range(len(self.indexes)):
+            index = self.indexes[i]
+            if not isinstance(index, (paddle.fluid.core.VarBase,
+                                      paddle.fluid.Variable)):  #tensor
+                raise ValueError(
+                    "only support list/tensor index, but received {}.".format(
+                        type(index)))
+
+        # if len(self.indexes)>1:
```
Comment (zyfncg) [inline python/paddle/fluid/variable_index.py]: delete?

File: python/paddle/fluid/variable_index.py [hunk 16]
```diff
@@ -16,10 +16,177 @@
...
+        for i in range(len(shape) - 2, -1, -1):
+            s[i] = shape[i + 1] * s[i + 1]
+
+        return s
+
+    def numel(self, shape):
+        return reduce(lambda x, y: x * y, shape)
+
+    def get_offset_stride(self, tensor_shape):
+        for i in range(len(self.indexes)):
```
Comment (zyfncg) [inline python/paddle/fluid/variable_index.py]: use `index in self.indexes` derectly?

File: python/paddle/fluid/tests/unittests/test_variable.py [hunk 17]
```diff
@@ -472,5 +478,455 @@ def test_slice(self):
...
+
+    def test_dygraph_list_index(self):
+        paddle.disable_static()
+
+        inps_shape = [3, 4, 5, 3]
+        array = np.arange(self.numel(inps_shape)).reshape(inps_shape)
+
+        index_shape = [2, 3, 4, 5, 6]
+        index = np.arange(self.numel(index_shape)).reshape(index_shape)
+        for i in range(len(inps_shape) - 1):
```
Comment (zyfncg) [inline python/paddle/fluid/tests/unittests/test_variable.py]: Are test cases same in the loop?

File: python/paddle/fluid/tests/unittests/test_variable.py [hunk 18]
```diff
@@ -472,5 +478,455 @@ def test_slice(self):
...
+        paddle.enable_static()
+
+        inps_shape = [3, 4, 5, 2]
+        array = np.arange(
+            self.numel(inps_shape), dtype='float32').reshape(inps_shape)
+
+        index_shape = [3, 3, 2, 1]
+        index = np.arange(self.numel(index_shape)).reshape(index_shape)
+
+        for i in range(3):
```
Comment (zyfncg) [inline python/paddle/fluid/tests/unittests/test_variable.py]: Are test cases same in the loop?

File: python/paddle/fluid/tests/unittests/test_variable.py [hunk 19]
```diff
@@ -472,5 +478,455 @@ def test_slice(self):
...
+            self.numel(inps_shape), dtype='float32').reshape(inps_shape)
+
+        index_shape = [3, 3, 1, 2]
+        index = np.arange(self.numel(index_shape)).reshape(index_shape)
+
+        value_shape = inps_shape[3:]
+        value_np = np.arange(
+            self.numel(value_shape), dtype='float32').reshape(value_shape) + 100
+
+        for i in range(3):
```
Comment (zyfncg) [inline python/paddle/fluid/tests/unittests/test_variable.py]: same above

File: python/paddle/fluid/tests/unittests/test_variable.py [hunk 20]
```diff
@@ -472,5 +478,455 @@ def test_slice(self):
...
+            self.numel(inps_shape), dtype='float32').reshape(inps_shape)
+
+        index_shape = [4, 3, 2, 2]
+        index = np.arange(self.numel(index_shape)).reshape(index_shape)
+
+        value_shape = [3]
+        value_np = np.arange(
+            self.numel(value_shape), dtype='float32').reshape(value_shape) + 100
+
+        for i in range(4):
```
Comment (zyfncg) [inline python/paddle/fluid/tests/unittests/test_variable.py]: same above

File: python/paddle/fluid/tests/unittests/test_variable.py [hunk 21]
```diff
@@ -472,5 +478,455 @@ def test_slice(self):
...
+        index_shape = [2, 3, 4]
+        index1 = np.arange(
+            self.numel(index_shape), dtype='int32').reshape(index_shape)
+        index2 = np.arange(
+            self.numel(index_shape), dtype='int32').reshape(index_shape) + 2
+
+        value_shape = [4]
+        value_np = np.arange(
+            self.numel(value_shape), dtype='float32').reshape(value_shape) + 100
+        for zz_ in range(3):
```
Comment (zyfncg) [inline python/paddle/fluid/tests/unittests/test_variable.py]: same above

File: python/paddle/fluid/tests/unittests/test_variable.py [hunk 22]
```diff
@@ -472,5 +478,455 @@ def test_slice(self):
...
+        array = np.arange(
+            self.numel(inps_shape), dtype='float32').reshape(inps_shape)
+
+        index_shape = [2, 3, 4]
+        index1 = np.arange(
+            self.numel(index_shape), dtype='int32').reshape(index_shape)
+        index2 = np.arange(
+            self.numel(index_shape), dtype='int32').reshape(index_shape) + 2
+
+        for zz_ in range(3):
```
Comment (zyfncg) [inline python/paddle/fluid/tests/unittests/test_variable.py]: same above

File: python/paddle/fluid/tests/unittests/test_variable.py [hunk 23]
```diff
@@ -472,5 +478,455 @@ def test_slice(self):
...
+        inps_shape = [3, 4, 5, 4]
+        array = np.arange(
+            self.numel(inps_shape), dtype='float32').reshape(inps_shape)
+        index_shape = [2, 3, 4]
+        index1 = np.arange(
+            self.numel(index_shape), dtype='int32').reshape(index_shape)
+        index2 = np.arange(
+            self.numel(index_shape), dtype='int32').reshape(index_shape) + 2
+
+        for zz_ in range(3):
```
Comment (zyfncg) [inline python/paddle/fluid/tests/unittests/test_variable.py]: same above

File: python/paddle/fluid/dygraph/varbase_patch_methods.py (hunk 10 above)
Comment (hbwx24) [reply python/paddle/fluid/dygraph/varbase_patch_methods.py]: Done, thx.

File: python/paddle/fluid/variable_index.py (hunk 12 above)
Comment (hbwx24) [reply python/paddle/fluid/variable_index.py]: Done, thx.

File: python/paddle/fluid/variable_index.py (hunk 13 above)
Comment (hbwx24) [reply python/paddle/fluid/variable_index.py]: Done, thx.

File: python/paddle/fluid/variable_index.py (hunk 14 above)
Comment (hbwx24) [reply python/paddle/fluid/variable_index.py]: Forgot to delete the comment. Has deleted it.

File: python/paddle/fluid/dygraph/varbase_patch_methods.py (hunk 11 above)
Comment (hbwx24) [reply python/paddle/fluid/dygraph/varbase_patch_methods.py]: Done, thx.

File: python/paddle/fluid/variable_index.py (hunk 15 above)
Comment (hbwx24) [reply python/paddle/fluid/variable_index.py]: Done, thx.

File: python/paddle/fluid/variable_index.py (hunk 16 above)
Comment (hbwx24) [reply python/paddle/fluid/variable_index.py]: Done, thx.

File: python/paddle/fluid/tests/unittests/test_variable.py (hunk 17 above)
Comment (hbwx24) [reply python/paddle/fluid/tests/unittests/test_variable.py]: Done, thx.

File: python/paddle/fluid/tests/unittests/test_variable.py (hunk 18 above)
Comment (hbwx24) [reply python/paddle/fluid/tests/unittests/test_variable.py]: Done, thx.

File: python/paddle/fluid/tests/unittests/test_variable.py (hunk 19 above)
Comment (hbwx24) [reply python/paddle/fluid/tests/unittests/test_variable.py]: Done, thx.

File: python/paddle/fluid/tests/unittests/test_variable.py (hunk 20 above)
Comment (hbwx24) [reply python/paddle/fluid/tests/unittests/test_variable.py]: Done, thx.

File: python/paddle/fluid/tests/unittests/test_variable.py (hunk 21 above)
Comment (hbwx24) [reply python/paddle/fluid/tests/unittests/test_variable.py]: Done, thx.

File: python/paddle/fluid/tests/unittests/test_variable.py (hunk 22 above)
Comment (hbwx24) [reply python/paddle/fluid/tests/unittests/test_variable.py]: Done, thx.

File: python/paddle/fluid/tests/unittests/test_variable.py (hunk 23 above)
Comment (hbwx24) [reply python/paddle/fluid/tests/unittests/test_variable.py]: Done, thx.

**Review by chenwhql — APPROVED** [review]
LGTM，动态图现在复用太多Python逻辑，后续需要解决由此引入的性能问题

File: paddle/fluid/operators/index_select_op.cc (hunk 1 above)
Comment (hbwx24) [reply paddle/fluid/operators/index_select_op.cc]: > LGTM，动态图现在复用太多Python逻辑，后续需要解决由此引入的性能问题

好的

**Pull request merged**